rat/compiler.py

```diff
@@ -133,362 +384,190 @@ def _identify_primary_symbols(self):
         """
+        self.symbol_table = SymbolTable(self.data_df)
 
-        # Make sure that all secondary parameter uses precede primary uses (or throw an error)
-        N_lines = len(self.expr_tree_list)
-        for i, top_expr in enumerate(self.expr_tree_list):
-            primary_symbol = self._get_primary_symbol_from_statement(top_expr)
-
-            primary_key = primary_symbol.get_key()
-
-            # There aren't any restrictions on primary/secondary data
-            if isinstance(primary_symbol, ops.Data):
-                continue
+        for top_expr in self.expr_tree_list:
+            if isinstance(top_expr, ast.Distr):
+                lhs_variable_key = top_expr.variate.get_key()
+            elif isinstance(top_expr, ast.Assignment):
+                lhs_variable_key = top_expr.lhs.get_key()
+
+            primary_variable = self._get_primary_symbol_from_statement(top_expr)
+            primary_variable_key = primary_variable.get_key()
+
+            if isinstance(primary_variable, ast.Data):
+                self.symbol_table.upsert(variable_name=primary_variable.name, variable_type=VariableType.DATA)
+            elif isinstance(primary_variable, ast.Param):
+                self.symbol_table.upsert(variable_name=primary_variable.name, variable_type=VariableType.PARAM)
+
+            subscript_aliases: Dict[str, Set] = {}
+            # this dict stores any subscript aliases of the primary dataframe.
+
+            # find the primary variable and its subscript aliases
+            match primary_variable:
+                case ast.Data():
+                    # these are the allowed subscript names in scope
+                    allowed_subscript_names = tuple(self.data_df.columns)
+                case ast.Param():
+                    try:
+                        primary_variable_record = self.symbol_table.lookup(primary_variable_key)
+                    except KeyError:
+                        msg = f"Primary variable {primary_variable_key} does not exist on the symbol table. Did you define the variable before using it as a primary variable?"
+                        raise CompileError(msg, self.model_code_string, primary_variable.line_index, primary_variable.column_index)
+                    else:
+                        if primary_variable_record.subscript_length > 0:
+                            # these are the allowed subscript names in scope
 
-            for j in range(i + 1, N_lines):
-                for secondary_symbol in ops.search_tree(self.expr_tree_list[j], ops.Param):
-                    secondary_key = secondary_symbol.get_key()
+                            if primary_variable.subscript:
+                                for index, subscript_column in enumerate(primary_variable.subscript.names):
+                                    if subscript_column.name not in self.data_df_columns:
+                                        subscript_aliases[subscript_column.name] = {
```

Comment:
Okay. Aliases are almost the same as renames in the current compiler. Take the simple baseball model:
```
score_diff ~ normal(skill[home_team] - skill[away_team], sigma);
skill[team] ~ normal(0.0, tau);
```
Previously, for parameter `skill`, we'd identify that the base_df should consist of input columns `home_team` and `away_team`, renaming its generated base_df's column name to `team`.

In the new compiler, the subscript alias `skill` denotes the set `{home_team, away_team}`, which is also used as its base_df column name. We still do the renaming - but just keep track of the "true" input dataframe column names that the renamed(alias) column name is created from.

Reply:
Aaah, k, I want to go ahead and materialize the intermediate dataframes for all the variables (and have the subscripts appear so that they are not represented as an expression + the input dataframe).

Reasons being:
1. It's simpler in my head (if it isn't, tell meeeee)
2. I want multiple input dataframes
3. With multiple input dataframes there will be joins and complicated transformations and I don't think the expression + input tracking will be easy